### PR TITLE
[BigQuery] Supporting TIMESTAMP_NTZ

### PR DESCRIPTION
**Pull request by Tang8330**
As per title.

This including the encoding algo for DATETIME

### CODE REVIEW

File: clients/bigquery/storagewrite.go [hunk 1]
```diff
@@ -89,15 +91,45 @@ func columnsToMessageDescriptor(cols []columns.Column) (*protoreflect.MessageDes
...
+func encodePacked64DatetimeSeconds(dateTime time.Time) int64 {
+	var bitFieldDatetimeSeconds int64
+	bitFieldDatetimeSeconds |= int64(dateTime.Year() << yearShift)
+	bitFieldDatetimeSeconds |= int64(dateTime.Month() << monthShift)
+	bitFieldDatetimeSeconds |= int64(dateTime.Day() << dayShift)
+	bitFieldDatetimeSeconds |= int64(encodePacked32TimeSeconds(dateTime.UTC()))
+	return bitFieldDatetimeSeconds
+}
+
+// // This is a reimplementation of https://github.com/googleapis/java-bigquerystorage/blob/f79acb5cfdd12253bca1c41551c478400120d2f9/google-cloud-bigquerystorage/src/main/java/com/google/cloud/bigquery/storage/v1/CivilTimeEncoder.java#L248
```
Comment (nathan-artie) [inline clients/bigquery/storagewrite.go:130]: Extra `//`

File: clients/bigquery/storagewrite.go [hunk 2]
```diff
@@ -89,15 +91,45 @@ func columnsToMessageDescriptor(cols []columns.Column) (*protoreflect.MessageDes
...
+// This is a reimplementation of https://github.com/googleapis/java-bigquerystorage/blob/f79acb5cfdd12253bca1c41551c478400120d2f9/google-cloud-bigquerystorage/src/main/java/com/google/cloud/bigquery/storage/v1/CivilTimeEncoder.java#L92
+func encodePacked32TimeSeconds(t time.Time) int32 {
+	var bitFieldTimeSeconds int32
+	bitFieldTimeSeconds |= int32(t.Hour()) << hourShift
+	bitFieldTimeSeconds |= int32(t.Minute()) << minuteShift
+	bitFieldTimeSeconds |= int32(t.Second()) << secondShift
+	return bitFieldTimeSeconds
+}
+
+// // This is a reimplementation of https://github.com/googleapis/java-bigquerystorage/blob/f79acb5cfdd12253bca1c41551c478400120d2f9/google-cloud-bigquerystorage/src/main/java/com/google/cloud/bigquery/storage/v1/CivilTimeEncoder.java#L187
```
Comment (nathan-artie) [inline clients/bigquery/storagewrite.go:120]: Extra `//`.

File: clients/bigquery/storagewrite.go [hunk 3]
```diff
@@ -89,15 +91,45 @@ func columnsToMessageDescriptor(cols []columns.Column) (*protoreflect.MessageDes
...
+	bitFieldDatetimeSeconds |= int64(dateTime.Year() << yearShift)
+	bitFieldDatetimeSeconds |= int64(dateTime.Month() << monthShift)
+	bitFieldDatetimeSeconds |= int64(dateTime.Day() << dayShift)
+	bitFieldDatetimeSeconds |= int64(encodePacked32TimeSeconds(dateTime.UTC()))
+	return bitFieldDatetimeSeconds
+}
+
+// // This is a reimplementation of https://github.com/googleapis/java-bigquerystorage/blob/f79acb5cfdd12253bca1c41551c478400120d2f9/google-cloud-bigquerystorage/src/main/java/com/google/cloud/bigquery/storage/v1/CivilTimeEncoder.java#L248
+func encodePacked64DatetimeMicros(dateTime time.Time) int64 {
+	return encodePacked64DatetimeSeconds(dateTime)<<microLength | int64(dateTime.Nanosecond()/1_000)
```
Comment (nathan-artie) [inline clients/bigquery/storagewrite.go:132]: Nit: inconsistant with `/1000` above.

File: clients/bigquery/storagewrite_test.go [hunk 4]
```diff
@@ -101,6 +101,58 @@ func TestEncodePacked64TimeMicros(t *testing.T) {
...
+
+	assert.Equal(t, int32(0), encodePacked32TimeSeconds(epoch))
+	assert.Equal(t, int32(1), encodePacked32TimeSeconds(epoch.Add(time.Duration(1)*time.Second)))
+	assert.Equal(t, int32(1<<6), encodePacked32TimeSeconds(epoch.Add(time.Duration(1)*time.Minute)))
+	assert.Equal(t, int32(1<<12), encodePacked32TimeSeconds(epoch.Add(time.Duration(1)*time.Hour)))
+	assert.Equal(t, int32(1<<12+1), encodePacked32TimeSeconds(epoch.Add(time.Duration(1)*time.Hour+time.Duration(1)*time.Second)))
+}
+
+func TestEncodePacked64DatetimeSeconds(t *testing.T) {
+	_ts := time.Date(2024, 10, 24, 13, 1, 2, 3000000, time.UTC)
```
Comment (nathan-artie) [inline clients/bigquery/storagewrite_test.go:115]: Nit: Do these need to be prefixed with an underscore?

**Review by nathan-artie — APPROVED** [review]

File: clients/bigquery/storagewrite_test.go (hunk 4 above)
Comment (Tang8330) [reply clients/bigquery/storagewrite_test.go:115]: Nope, can change

File: clients/bigquery/storagewrite.go (hunk 3 above)
Comment (Tang8330) [reply clients/bigquery/storagewrite.go:132]: sg

**Pull request merged**